daemon/containerd/image_list.go

```diff
@@ -199,14 +221,39 @@ func (i *ImageService) singlePlatformImage(ctx context.Context, contentStore con
 			repoTags = append(repoTags, rawImg.Name)
 		}
 	} else {
-		repoTags = append(repoTags, reference.TagNameOnly(ref).String())
-
 		digested, err := reference.WithDigest(reference.TrimNamed(ref), target)
 		if err != nil {
 			logger.WithError(err).Error("failed to create digested reference")
 		} else {
 			repoDigests = append(repoDigests, digested.String())
 		}
+
+		for _, imageRef := range references {
+			if isDanglingImage(imageRef) {
+				continue
+			}
+			if !filter(imageRef) {
+				continue
+			}
+
+			ref, err := reference.ParseNormalizedNamed(imageRef.Name)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			if image.RealTarget.Digest.String() == imageRef.Target.Digest.String() {
+				repoTags = append(repoTags, reference.FamiliarString(ref))
+			}
+		}
```

Comment:
performance nit: We can easily avoid iterating over all images filtering them over and over again in every `singlePlatformImage` call.

See: https://github.com/vvoland/moby/commit/d2729875afe1a907cc4a0d90aa7aeda3780be71d

Reply:
Nice, doesn't look like it works for dangling images at all?

Reply:
I'll fix it and take your perf nit